### PR TITLE
Validate Sampled=1 for Vulkan ImageQuerySizeLod, ImageQueryLevels, ImageQueryLod

### LINKED ISSUE #4102: validate VUID-StandaloneSpirv-OpImageQuerySizeLod-04659
VUID-StandaloneSpirv-OpImageQuerySizeLod-04659
OpImageQuerySizeLod, and OpImageQueryLevels must only consume an “Image” operand whose type has its “Sampled” operand set to 1

It must be guarded by being in the Vulkan environment.

cc: @sfricke-samsung in case you're trying to do this too somewhere.

### PR DESCRIPTION
**Pull request by dneto0**
Fixes #4102

### CODE REVIEW

File: source/val/validate_image.cpp [hunk 1]
```diff
@@ -1903,6 +1911,15 @@ spv_result_t ValidateImageQueryLod(ValidationState_t& _,
            << "Expected Coordinate to have at least " << min_coord_size
            << " components, but given only " << actual_coord_size;
   }
+
+  const auto target_env = _.context()->target_env;
```
Comment (alan-baker) [inline source/val/validate_image.cpp:1915]: This seems off. The `SampledImage` operand must be an `OpTypeSampledImage`. The validator check that `OpSampledImage` must have `Sampled` equal to 1 for Vulkan, but does not check that for `OpTypeSampledImage`. I would suggest removing this check and adding one in `ValidateTypeSampleImage()`.

File: source/val/validate_image.cpp [hunk 2]
```diff
@@ -1729,6 +1729,14 @@ spv_result_t ValidateImageQuerySizeLod(ValidationState_t& _,
     return _.diag(SPV_ERROR_INVALID_DATA, inst) << "Image 'MS' must be 0";
   }
 
+  const auto target_env = _.context()->target_env;
+  if (spvIsVulkanEnv(target_env)) {
+    if (info.sampled != 1) {
+      return _.diag(SPV_ERROR_INVALID_DATA, inst)
```
Comment (alan-baker) [inline source/val/validate_image.cpp:1735]: Probably worth adding the Vulkan VUID for this error (and using it's language). VUID is VUID-StandaloneSpirv-OpImageQuerySizeLod-04659. Change the error to:
> OpImageQuerySizeLod must only consume an “Image” operand whose type has its “Sampled” operand set to 1

File: source/val/validate_image.cpp [hunk 3]
```diff
@@ -1939,6 +1956,14 @@ spv_result_t ValidateImageQueryLevelsOrSamples(ValidationState_t& _,
       return _.diag(SPV_ERROR_INVALID_DATA, inst)
              << "Image 'Dim' must be 1D, 2D, 3D or Cube";
     }
+    const auto target_env = _.context()->target_env;
+    if (spvIsVulkanEnv(target_env)) {
```
Comment (alan-baker) [inline source/val/validate_image.cpp:1960]: Same as above, use the VUID and language from Vulkan:
>OpImageQueryLevels must only consume an “Image” operand whose type has its “Sampled” operand set to 1

File: test/val/val_image_test.cpp [hunk 4]
```diff
@@ -4257,6 +4366,102 @@ OpExecutionMode %main DerivativeGroupLinearNV
...
+%voidfn = OpTypeFunction %void
+
+; Test with an image with Sampled = 2
+; In Vulkan it Sampled must be 1 or 2, checked in another part of the
+; validation flow.
+%type_image_f32_2d_0002 = OpTypeImage %f32 2D 0 0 0 2 Rgba32f
+
+; TODO(dneto): It seems wrong that we should be able to form a sampled image type
+; with an image type parameter that uses Sampled=2.
+%sampled_image_ty = OpTypeSampledImage %type_image_f32_2d_0002
```
Comment (alan-baker) [inline test/val/val_image_test.cpp:4445]: This should be a validation error.

**Review by alan-baker — CHANGES_REQUESTED** [review]

File: test/val/val_image_test.cpp (hunk 4 above)
Comment (dneto0) [reply test/val/val_image_test.cpp:4445]: I filed Khronos internal SPIR-V issue 631 for a spec update.
I'm happy to implement the extra validation.

File: source/val/validate_image.cpp (hunk 1 above)
Comment (dneto0) [reply source/val/validate_image.cpp]: Done.
I'll check agianst Vulkan CTS master branch.

File: source/val/validate_image.cpp [hunk 5]
```diff
@@ -1729,6 +1743,16 @@ spv_result_t ValidateImageQuerySizeLod(ValidationState_t& _,
     return _.diag(SPV_ERROR_INVALID_DATA, inst) << "Image 'MS' must be 0";
   }
 
+  const auto target_env = _.context()->target_env;
+  if (spvIsVulkanEnv(target_env)) {
+    if (info.sampled != 1) {
+      return _.diag(SPV_ERROR_INVALID_DATA, inst)
+             << _.VkErrorID(4659)
```
Comment (alan-baker) [inline source/val/validate_image.cpp:1750]: It seems you missed adding the entry in ValidationState.cpp

**Review by alan-baker — CHANGES_REQUESTED** [review]

File: source/val/validate_image.cpp (hunk 5 above)
Comment (dneto0) [reply source/val/validate_image.cpp:1750]: Done.  Thanks!

**Comment by dneto0** [comment]
I tested this patch against the Vulkan CTS vk-build-programs.

```
Writing test log into TestResults.qpa
SPIR-V versions: baseline: 1.0, max supported: 1.5
DONE: 1588403 passed, 0 failed, 0 not supported
```

**Review by alan-baker — APPROVED** [review]

**Comment by dneto0** [comment]
Rebased against master

**Pull request merged**